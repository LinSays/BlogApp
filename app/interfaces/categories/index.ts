import { Categories } from "./categories.interface";

export { Categories };
