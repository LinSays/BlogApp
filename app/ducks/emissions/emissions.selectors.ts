/* eslint-disable  @typescript-eslint/explicit-module-boundary-types */
import { filter, propEq, find, pathOr, pipe } from "ramda";

import { EmissionType } from "interfaces";

import { namespace } from "./emissions.slice";

const getAllEmissions = pathOr([], [namespace]);

const getEmissionById = (state, id: string) => find(propEq("id", id))(state[namespace]);

const getEmissionsToMitigate = (state) =>
  pipe(getAllEmissions, filter(propEq("isMitigated", false)))(state);

const getEmissionsMitigated = (state) =>
  pipe(getAllEmissions, filter(propEq("isMitigated", true)))(state);

const isTransportEmission = (emission) => emission.emissionType === EmissionType.transport;

const isFoodEmission = (emission) => emission.emissionType === EmissionType.food;

const isPurchaseEmission = (emission) => emission.emissionType === EmissionType.purchase;

const isFashionEmission = (emission) => emission.emissionType === EmissionType.fashion;

const isStreamingEmission = (emission) => emission.emissionType === EmissionType.streaming;

const isElectricityEmission = (emission) => emission.emissionType === EmissionType.electricity;

const isCustomEmission = (emission) => emission.emissionType === EmissionType.custom;

const isMealEmission = (emission) => emission.emissionType === EmissionType.meal;

const isEatableEmission = (emission) =>
  emission.emissionType === EmissionType.meal || emission.emissionType === EmissionType.food;

const isOtherEmission = (emission) =>
  emission.emissionType === EmissionType.custom ||
  emission.emissionType === EmissionType.streaming ||
  emission.emissionType === EmissionType.electricity ||
  emission.emissionType === EmissionType.fashion ||
  emission.emissionType === EmissionType.purchase ||
  emission.emissionType === EmissionType.meal;

const getTransportEmissions = pipe(getAllEmissions, filter(isTransportEmission));

const getFoodEmissions = pipe(getAllEmissions, filter(isFoodEmission));

const getStreamingEmissions = pipe(getAllEmissions, filter(isStreamingEmission));

const getElectricityEmissions = pipe(getAllEmissions, filter(isElectricityEmission));

const getPurchaseEmissions = pipe(getAllEmissions, filter(isPurchaseEmission));

const getFashionEmissions = pipe(getAllEmissions, filter(isFashionEmission));

const getMealEmissions = pipe(getAllEmissions, filter(isMealEmission));

const getEatableEmissions = pipe(getAllEmissions, filter(isEatableEmission));

const getCustomEmissions = pipe(getAllEmissions, filter(isCustomEmission));

const getOtherEmissions = pipe(getAllEmissions, filter(isOtherEmission));

export default {
  getAllEmissions,
  getTransportEmissions,
  getFoodEmissions,
  getStreamingEmissions,
  getElectricityEmissions,
  getFashionEmissions,
  getPurchaseEmissions,
  getCustomEmissions,
  getOtherEmissions,
  getMealEmissions,
  getEatableEmissions,
  getEmissionById,
  getEmissionsToMitigate,
  getEmissionsMitigated,
};
