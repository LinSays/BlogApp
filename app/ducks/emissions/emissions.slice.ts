import { createSlice, PayloadAction } from "@reduxjs/toolkit";

import { Emission } from "interfaces";

const initialState: Emission[] = [];

const emissions = createSlice({
  name: "emissions",
  initialState,
  reducers: {
    createEmission(state, action: PayloadAction<Emission>) {
      state.push(action.payload);
    },
    deleteEmission(state, action: PayloadAction<string>) {
      return (state = state.filter((item) => item.id !== action.payload));
    },
    toggleIsMitigated(state, action: PayloadAction<string>) {
      return (state = state.map((item) =>
        item.id === action.payload ? { ...item, isMitigated: !item.isMitigated } : item
      ));
    },
    deleteAllEmissions() {
      return initialState;
    },
    loadEmissionsFromImport(state, action: PayloadAction<Array<Emission>>) {
      return (state = action.payload);
    },
  },
});

const {
  createEmission,
  deleteEmission,
  toggleIsMitigated,
  deleteAllEmissions,
  loadEmissionsFromImport,
} = emissions.actions;

export const actions = {
  createEmission,
  deleteEmission,
  toggleIsMitigated,
  deleteAllEmissions,
  loadEmissionsFromImport,
};

export const namespace = emissions.name;

export const reducer = emissions.reducer;
