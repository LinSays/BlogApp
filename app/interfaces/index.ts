import { Categories } from "./categories";
import NavStatelessComponent from "./navigation";

export { NavStatelessComponent, Categories };
