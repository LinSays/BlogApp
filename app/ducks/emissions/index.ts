import { actions, namespace, reducer } from "./emissions.slice";
import selectors from "./emissions.selectors";

export default { actions, namespace, reducer, selectors };
