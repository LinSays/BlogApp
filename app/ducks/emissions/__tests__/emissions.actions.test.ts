import { TransportType } from "carbon-footprint";

import { Emission, EmissionType } from "interfaces";

import emissions from "../";

describe("Emissions actions should", () => {
  it("export expected actions", () => expect(emissions.actions).toMatchSnapshot());

  it("be able to create an emission", () => {
    const emission: Emission = {
      id: "123",
      creationDate: "2020-01-26T11:04:55.334Z",
      emissionType: EmissionType.custom,
      emissionModelType: TransportType.boat,
      value: 200,
      isMitigated: false,
    };

    const expectedAction = {
      type: emissions.actions.createEmission.toString(),
      payload: emission,
    };

    expect(emissions.actions.createEmission(emission)).toEqual(expectedAction);
  });

  it("be able to delete an emission", () => {
    const expectedAction = {
      type: emissions.actions.deleteEmission.toString(),
      payload: "123",
    };

    expect(emissions.actions.deleteEmission("123")).toEqual(expectedAction);
  });

  it("be able to toggle isMitigated property of an emission", () => {
    const expectedAction = {
      type: emissions.actions.toggleIsMitigated.toString(),
      payload: "123",
    };

    expect(emissions.actions.toggleIsMitigated("123")).toEqual(expectedAction);
  });

  it("be able to load emissions", () => {
    const emissionsList: Array<Emission> = [
      {
        id: "1",
        creationDate: "2020-01-26T11:04:55.334Z",
        emissionType: EmissionType.transport,
        emissionModelType: TransportType.boat,
        value: 200,
        isMitigated: false,
      },
      {
        id: "2",
        creationDate: "2020-02-26T11:04:55.334Z",
        emissionType: EmissionType.transport,
        emissionModelType: TransportType.bus,
        value: 100,
        isMitigated: false,
      },
    ];

    const expectedAction = {
      type: emissions.actions.loadEmissionsFromImport.toString(),
      payload: emissionsList,
    };

    expect(emissions.actions.loadEmissionsFromImport(emissionsList)).toEqual(expectedAction);
  });

  it("be able to delete all emissions", () => {
    const expectedAction = {
      type: emissions.actions.deleteAllEmissions.toString(),
      payload: undefined,
    };

    expect(emissions.actions.deleteAllEmissions()).toEqual(expectedAction);
  });
});
