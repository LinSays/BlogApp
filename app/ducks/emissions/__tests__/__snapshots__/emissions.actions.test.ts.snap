// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Emissions actions should export expected actions 1`] = `
Object {
  "createEmission": [Function],
  "deleteAllEmissions": [Function],
  "deleteEmission": [Function],
  "loadEmissionsFromImport": [Function],
  "toggleIsMitigated": [Function],
}
`;
